### PR TITLE
Documentation for the linear regression

### PR DESCRIPTION
**Pull request by George3d6**

### CODE REVIEW

File: lightwood/mixer/regression.py [hunk 1]
```diff
@@ -41,10 +57,24 @@ def fit(self, train_data: EncodedDs, dev_data: EncodedDs) -> None:
         log.info(f'Regression based correlation of: {self.model.score(X, Y)}')
 
     def partial_fit(self, train_data: EncodedDs, dev_data: EncodedDs) -> None:
+        """
+        Fits the Neural mixer on some data, this refits the model entirely rather than updating it
```
Comment (paxcema) [inline lightwood/mixer/regression.py:61]: Typo: 
> Fits the _regression_ mixer

File: lightwood/mixer/regression.py [hunk 2]
```diff
@@ -12,11 +12,21 @@
...
+    A mixer that runs a simple linear regression using the (Encoded) features to predict the target.
+    Supports all types because they are all encoded numerically.
+    """ # noqa
     model: LinearRegression
     label_map: dict
     supports_proba: bool
 
     def __init__(self, stop_after: int, target_encoder: BaseEncoder, dtype_dict: dict, target: str):
+        """
+        :param stop_after: Maximum amount of time it should train for, currently ignored
```
Comment (hakunanatasha) [inline lightwood/mixer/regression.py:31]: in seconds? minutes? add the time interval.

File: lightwood/mixer/regression.py [hunk 3]
```diff
@@ -25,6 +41,12 @@ def __init__(self, stop_after: int, target_encoder: BaseEncoder, dtype_dict: dic
         self.stable = False
 
     def fit(self, train_data: EncodedDs, dev_data: EncodedDs) -> None:
+        """
+        Fits `LinearRegression` model on input feature data to provide predictions.
+
+        :param train_data: The EncodedDs on which to train the network
```
Comment (hakunanatasha) [inline lightwood/mixer/regression.py:47]: what network?

File: lightwood/mixer/regression.py [hunk 4]
```diff
@@ -25,6 +41,12 @@ def __init__(self, stop_after: int, target_encoder: BaseEncoder, dtype_dict: dic
...
+        :param dev_data: Data used for early stopping and hyperparameter determination
```
Comment (hakunanatasha) [inline lightwood/mixer/regression.py:48]: for linear regression, there are no hyper parameters or early stopping

File: lightwood/mixer/regression.py [hunk 5]
```diff
@@ -41,10 +63,24 @@ def fit(self, train_data: EncodedDs, dev_data: EncodedDs) -> None:
         log.info(f'Regression based correlation of: {self.model.score(X, Y)}')
 
     def partial_fit(self, train_data: EncodedDs, dev_data: EncodedDs) -> None:
+        """
+        Fits the Neural mixer on some data, this refits the model entirely rather than updating it
+
+        :param train_data: The EncodedDs on which to train the network
```
Comment (hakunanatasha) [inline lightwood/mixer/regression.py:69]: there is no network. this is the simplest model.

File: lightwood/mixer/regression.py [hunk 6]
```diff
@@ -41,10 +63,24 @@ def fit(self, train_data: EncodedDs, dev_data: EncodedDs) -> None:
...
+        :param dev_data: Data used for early stopping and hyperparameter determination
```
Comment (hakunanatasha) [inline lightwood/mixer/regression.py:70]: likewise no network, no hyperparams, no early stopping

File: lightwood/mixer/regression.py [hunk 7]
```diff
@@ -41,10 +63,24 @@ def fit(self, train_data: EncodedDs, dev_data: EncodedDs) -> None:
         log.info(f'Regression based correlation of: {self.model.score(X, Y)}')
 
     def partial_fit(self, train_data: EncodedDs, dev_data: EncodedDs) -> None:
```
Comment (hakunanatasha) [inline lightwood/mixer/regression.py:65]: The inputs to partial-fit are incorrectly named. We should change it to `old_data` and `new_data`. I suspect this is a holdover of train's inputs, where dev data is specifically reserved for hyperparam/earlystopping etc.

In this case, the name "dev_data" is actually used to adjust the model with new information; this is *unintuitive*. We need to fix this for the entire codebase.

File: lightwood/mixer/regression.py [hunk 8]
```diff
@@ -41,10 +63,24 @@ def fit(self, train_data: EncodedDs, dev_data: EncodedDs) -> None:
...
+
+        :param train_data: The EncodedDs on which to train the network
+        :param dev_data: Data used for early stopping and hyperparameter determination
+        """
         self.fit(train_data, dev_data)
 
     def __call__(self, ds: EncodedDs,
                  args: PredictionArguments = PredictionArguments()) -> pd.DataFrame:
+        """
+        Make predictions based on datasource similar to the one used to fit (sans the target column)
```
Comment (hakunanatasha) [inline lightwood/mixer/regression.py:77]: This doesn't even make sense, what do you mean similar? The same data structure? This should be obvious.

File: lightwood/mixer/regression.py [hunk 9]
```diff
@@ -41,10 +63,24 @@ def fit(self, train_data: EncodedDs, dev_data: EncodedDs) -> None:
...
+
+        :param ds: The EncodedDs for which to generate the predictions
```
Comment (hakunanatasha) [inline lightwood/mixer/regression.py:79]: Sphinx already type-hints you don't need to call it by the dataclass.

Comment (hakunanatasha) [reply lightwood/mixer/regression.py:79]: ":param ds: The featurized dataset used to generate predictions"

File: lightwood/mixer/regression.py [hunk 10]
```diff
@@ -41,10 +63,24 @@ def fit(self, train_data: EncodedDs, dev_data: EncodedDs) -> None:
...
+        :param arg: Argument for predicting
```
Comment (hakunanatasha) [inline lightwood/mixer/regression.py:80]: ":param arg: Any additional arguments used in predicting"

Comment (hakunanatasha) [reply lightwood/mixer/regression.py:80]: does this even have additional args? I don't know if this is just there to make things modular to the mixer format, but ideally we don't even need to have this extra term.

File: lightwood/mixer/regression.py (hunk 1 above)
Comment (hakunanatasha) [reply lightwood/mixer/regression.py:67]: You need to actually discuss how partial-fit is handled in LinearRegression. LinearRegression is cheap, hence we simply re-fit the model with more data (old data + new data). This is NOT the same as in other models, for example, neural networks, where you can get away with training on just new data assuming the old data representations have been "learned" or fine-tuning.

**Review by hakunanatasha — CHANGES_REQUESTED** [review]
These functions are incorrectly annotated (I assume because they are borrowed from other areas). Please make sure to detail arguments and actually provide a deeper reference as to what's going on.

File: lightwood/mixer/regression.py (hunk 2 above)
Comment (George3d6) [reply lightwood/mixer/regression.py:31]: Oh, good point. This is an issue in like half a dozen places (we don't specify) the unit. So I went ahead and did a more generic fix where `stop_after` and `time_aim` are now typed as `seconds` so that it's obvious what they are all the time.

File: lightwood/mixer/regression.py (hunk 1 above)
Comment (hakunanatasha) [reply lightwood/mixer/regression.py]: There is no choice to motivate. If you retrain Linear regression with a new dataset, it's as if the old dataset was never seen. 

The problem is the term `partial_fit` is ambiguous, so we need to be clear what `refit` means, because in certain contexts `refitting` means that we can train without the old data (i.e. neural nets can be fine-tuned with new data). Terms like `update` and `refit` do not discuss *how* changes are made. It doesn't hurt to be explicit.

File: lightwood/api/types.py [hunk 11]
```diff
@@ -15,6 +15,9 @@
 import json
 
 
+seconds = float
+
+
 # See: https://www.python.org/dev/peps/pep-0589/ for how this works
 # Not very intuitive but very powerful abstraction, might be useful in other places (@TODO)
```
Comment (hakunanatasha) [inline lightwood/api/types.py:22]: why is this here? please remove it

**Review by hakunanatasha — CHANGES_REQUESTED** [review]
This is an over-engineered solution. When I said time interval, I merely mean to annotate in the docs "this is in seconds". Making a weird reserved type that is a float is pretty confusing, as people will then need to look around for what weird data type we've enforced `seconds` to be. It can be a float or int param argument, it's fine.

When we actually DO have a time-aim, the package we use to explicitly "stop" training has an expectation for the input. That's what we want to make clear.
